### PR TITLE
Explicit operation order determined by user definition

### PR DESCRIPTION
**Pull request by skim0119**
# Explicit operation order

- User-API would remain the same

## Description

Previous implementation has ambiguity in determining the order of operation for different boundary conditions: `forcing`, `contact`, `friction`, `connection`, etc. For example, `contact` and `friction` must be calculated after all other boundary conditions are in place, and strictly `contact` must be computed before `friction` since `friction` depends on `contact`. In various place in the code, we tried to rearange the order without explicitly letting user know: https://github.com/GazzolaLab/PyElastica/blob/87a2e74eb5cd1de0671847753efb630afaa1acf5/elastica/modules/base_system.py#L173-L182, https://github.com/GazzolaLab/PyElastica/blob/87a2e74eb5cd1de0671847753efb630afaa1acf5/elastica/modules/forcing.py#L72-L81.
This cause ambiguity on user, especially when user wants to define their own boundary condition.

## Solution

As long as we let users to write their own boundary conditions, we cannot anticipate every possible cases of operation dependencies. We are shifting to the policy that user must be aware of the boundary operation order.

This PR removes any "back-end" sorting the operation order. __Operation order is now strictly defined by the order of user inserting any boundary condition with `.using`.__ This applies to `forcing`, `contact`, and `connection` as of now.

## Some remaining TODOs:

- [x] Add warning messages on packaged `contact` and `friction` modules
- [x] Refactor synchronize operations: "first-in" boundary condition computes first
- [x] Test
  - [x] Fix existing tests
  - [x] Implement new test for testing operation order
- [x] Documentation ([f0fc544](https://github.com/GazzolaLab/PyElastica/pull/379/commits/f0fc544bb8be1458b03b41a5458ff453b04fea33))

### CODE REVIEW

File: elastica/modules/forcing.py [hunk 1]
```diff
@@ -54,37 +54,24 @@ def _finalize_forcing(self):
...
+            apply_forces = functools.partial(
+                forcing_instance.apply_forces, system=self._systems[sys_id]
+            )
+            apply_torques = functools.partial(
+                forcing_instance.apply_torques, system=self._systems[sys_id]
+            )
+
+            self._feature_group_synchronize.add_operators(
+                ext_force_torque, [apply_forces, apply_torques]
+            )
```
Comment (armantekinalp) [inline elastica/modules/forcing.py:71]: Don't we need to pass time here?

Comment (skim0119) [reply elastica/modules/forcing.py:71]: No, apply_force methods already include time as a parameter. Actually, we should add that parameter to connection and contact for consistency, which is why their implementation look a little different.

File: elastica/modules/connections.py [hunk 2]
```diff
@@ -60,49 +59,53 @@ def connect(
...
+                connect_instance.apply_torques(
+                    system_one=self._systems[first_sys_idx],
+                    index_one=first_connect_idx,
+                    system_two=self._systems[second_sys_idx],
+                    index_two=second_connect_idx,
+                )
+
+            self._feature_group_synchronize.add_operators(
+                connection, [apply_forces, apply_torques]
+            )
```
[the remark this answers is not in the document]
Comment (armantekinalp) [reply elastica/modules/connections.py:100]: Or is it because of time argument?

[the remark this answers is not in the document]
Comment (skim0119) [reply elastica/modules/connections.py:100]: #380 I added new issue, we'll clean out this part in later PR.

File: elastica/modules/contact.py [hunk 3]
```diff
@@ -61,30 +60,26 @@ def _finalize_contact(self) -> None:
...
                 self._systems[second_sys_idx],
             )
 
+            def apply_contact(time):
+                contact_instance.apply_contact(
+                    system_one=self._systems[first_sys_idx],
+                    system_two=self._systems[second_sys_idx],
+                )
+
+            self._feature_group_synchronize.add_operators(contact, [apply_contact])
```
[the remark this answers is not in the document]
Comment (skim0119) [reply elastica/modules/contact.py:78]: #380 I added new issue, we'll clean out this part in later PR.

File: elastica/modules/feature_group.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        for operator in operator_chain:
+            yield operator
+
+    def append_id(self, feature):
+        self._operator_ids.append(id(feature))
+        self._operator_collection.append([])
+
+    def add_operators(self, feature, operators: list[OperatorType]):
+        idx = self._operator_ids.index(id(feature))
+        self._operator_collection[idx].extend(operators)
```
[the remark this answers is not in the document]
Comment (skim0119) [reply elastica/modules/feature_group.py]: [c4ad237](https://github.com/GazzolaLab/PyElastica/pull/379/commits/c4ad237b784490611a26b62314749e56c584b0ed)
I added documentation and how to use. Also added unittest to have a script of "expected" usage.

File: elastica/modules/forcing.py [hunk 5]
```diff
@@ -54,37 +57,37 @@ def _finalize_forcing(self):
...
+                external_force_and_torque, [apply_forces, apply_torques]
+            )
+
+            self.warnings(external_force_and_torque)
+
+        self._ext_forces_torques = []
+        del self._ext_forces_torques
+
+    def warnings(self, external_force_and_torque):
+        from elastica.interaction import AnisotropicFrictionalPlane
```
Comment (armantekinalp) [inline elastica/modules/forcing.py:82]: We moved to friction plane under contact. Can you change accordingly.

File: elastica/modules/operator_group.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
+from elastica.typing import OperatorType
+
+from collections.abc import Iterable
+
+import itertools
+
+
+class OperatorGroupFIFO(Iterable):
```
Comment (armantekinalp) [inline elastica/modules/operator_group.py:8]: FIFO does it mean first initialized first operated? Can we Write long maybe?

File: elastica/modules/contact.py [hunk 7]
```diff
@@ -61,30 +63,36 @@ def _finalize_contact(self) -> None:
...
+
+            self.warnings(contact)
```
Comment (armantekinalp) [inline elastica/modules/contact.py:83]: smart way to check. Only issue, is what happens if we have rod-rod contact and frictional plane at the same time?

File: elastica/modules/operator_group.py (hunk 6 above)
Comment (skim0119) [reply elastica/modules/operator_group.py:8]: FIFO: first in first out
I think it is common-enough terminology for queuing

File: elastica/modules/contact.py (hunk 7 above)
Comment (skim0119) [reply elastica/modules/contact.py:83]: You'll get warning 😄

Comment (armantekinalp) [reply elastica/modules/contact.py:83]: Or two contacts.

File: elastica/modules/forcing.py (hunk 5 above)
Comment (skim0119) [reply elastica/modules/forcing.py:82]: I see `AnisotropicFrictionalPlane` in `interaction.py`:
https://github.com/GazzolaLab/PyElastica/blob/e71e54790d5b782ad1a2d9856888b09783251fd0/elastica/interaction.py#L158C7-L158C33

**Review by armantekinalp — APPROVED** [review]
LGTM

File: elastica/modules/forcing.py (hunk 5 above)
Comment (skim0119) [reply elastica/modules/forcing.py]: I removed warning for AnisotropicFrictionalPlane. @Ali-7800 will remove the class.
